rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -1781,11 +2462,91 @@ rmw_take_request(
   void * ros_request,
   bool * taken)
 {
-  static_cast<void>(service);
-  static_cast<void>(request_header);
-  static_cast<void>(ros_request);
-  static_cast<void>(taken);
-  return RMW_RET_UNSUPPORTED;
+  *taken = false;
+  RCUTILS_LOG_INFO_NAMED("rmw_zenoh_cpp", "[rmw_take_request]");
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(service, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(service->data, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(ros_request, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(taken, RMW_RET_INVALID_ARGUMENT);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    service,
+    service->implementation_identifier,
+    rmw_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    service->service_name, "service has no service name", RMW_RET_INVALID_ARGUMENT);
+
+  rmw_service_data_t * service_data = static_cast<rmw_service_data_t *>(service->data);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    service->data, "Unable to retrieve service_data from service", RMW_RET_INVALID_ARGUMENT);
+
+  z_owned_query_t query;
```

Comment:
It looks like we want to "move" the `z_owned_query_t` from `query_queue.front()` to `sequence_to_query_map` when some conditions are valid. I'm a bit concerned about creating this `z_owend_query_t object` and whether we need to drop it if the function returns before emplacing into the map?
Wondering if we should instantiate this as `z_owned_query_t * query` so that we won't have to drop anything and we can just `std::move(*query)` when emplacing into the map?

Reply:
Alternatively, the map and queue can store `unique_ptr<z_owned_query_t>` to make sure we're transferring ownership?

Reply:
> I'm a bit concerned about creating this `z_owend_query_t object` and whether we need to drop it if the function returns before emplacing into the map?

So I don't think this is actually a worry.  That's because we only `pop_front` the request if we totally succeeded in this function.  Thus, if we were to fail somewhere else in here, we would just leave it at the front of the replies.  That means we would reprocess it next time; that may or may not be what we want to do, but it is what we do now.

All of that said, we can make this a little clearer by using unique_ptr here, so I've gone ahead and done that (I'll push some fixes to your other PR).  Also I realized that the locking in here was completely bogus, so I'm going to fix that up as well.